File: swift-bridge/Sources/CoreDataBridge/PersistentContainer.swift
import CoreData
import Foundation

final class CDLoadTracker {
    private let lock = NSLock()
    private let expectedCount: Int
    private var completionCount = 0
    private var firstError: NSError?
    private let semaphore = DispatchSemaphore(value: 0)

    init(expectedCount: Int) {
        self.expectedCount = max(expectedCount, 1)
    }

    func record(error: NSError?) {
        lock.lock()
        defer { lock.unlock() }
        if firstError == nil {
            firstError = error
        }
        completionCount += 1
        if completionCount >= expectedCount {
            semaphore.signal()
        }
    }

    func wait(timeoutSeconds: Int32) -> NSError? {
        if semaphore.wait(timeout: .now() + .seconds(Int(timeoutSeconds))) == .timedOut {
            return cdBridgeNSError(code: CDR_TIMED_OUT, message: "Timed out waiting for loadPersistentStores")
        }
        return firstError
    }
}

@_cdecl("cd_persistent_store_description_new")
public func cdPersistentStoreDescriptionNew(
    _ path: UnsafePointer<CChar>?,
    _ outDescription: UnsafeMutablePointer<UnsafeMutableRawPointer?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    let description = cdURL(from: path).map(NSPersistentStoreDescription.init(url:)) ?? NSPersistentStoreDescription()
    outDescription?.pointee = cdRetain(description)
    return CDR_OK
}

@_cdecl("cd_persistent_store_description_get_type")
public func cdPersistentStoreDescriptionGetType(_ descriptionPtr: UnsafeMutableRawPointer?) -> UnsafeMutablePointer<CChar>? {
    guard let descriptionPtr else {
        return nil
    }
    let description: NSPersistentStoreDescription = cdBorrow(descriptionPtr)
    return cdCString(description.type)
}

@_cdecl("cd_persistent_store_description_set_type")
public func cdPersistentStoreDescriptionSetType(
    _ descriptionPtr: UnsafeMutableRawPointer?,
    _ storeType: UnsafePointer<CChar>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        guard let descriptionPtr, let storeType else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Missing store description or store type")
        }
        let description: NSPersistentStoreDescription = cdBorrow(descriptionPtr)
        description.type = String(cString: storeType)
        return CDR_OK
    } catch let error as NSError {
        cdWriteError(error, to: outError)
        return Int32(error.code)
    }
}

@_cdecl("cd_persistent_store_description_get_configuration")
public func cdPersistentStoreDescriptionGetConfiguration(_ descriptionPtr: UnsafeMutableRawPointer?) -> UnsafeMutablePointer<CChar>? {
    guard let descriptionPtr else {
        return nil
    }
    let description: NSPersistentStoreDescription = cdBorrow(descriptionPtr)
    return description.configuration.flatMap { cdCString($0) }
}

@_cdecl("cd_persistent_store_description_set_configuration")
public func cdPersistentStoreDescriptionSetConfiguration(
    _ descriptionPtr: UnsafeMutableRawPointer?,
    _ configuration: UnsafePointer<CChar>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        guard let descriptionPtr else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Missing store description")
        }
        let description: NSPersistentStoreDescription = cdBorrow(descriptionPtr)
        description.configuration = cdOptionalString(configuration)
        return CDR_OK
    } catch let error as NSError {
        cdWriteError(error, to: outError)
        return Int32(error.code)
    }
}

@_cdecl("cd_persistent_store_description_get_url")
public func cdPersistentStoreDescriptionGetURL(_ descriptionPtr: UnsafeMutableRawPointer?) -> UnsafeMutablePointer<CChar>? {
    guard let descriptionPtr else {
        return nil
    }
    let description: NSPersistentStoreDescription = cdBorrow(descriptionPtr)
    return description.url.flatMap { cdCString($0.path) }
}

@_cdecl("cd_persistent_store_description_set_url")
public func cdPersistentStoreDescriptionSetURL(
    _ descriptionPtr: UnsafeMutableRawPointer?,
    _ path: UnsafePointer<CChar>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        guard let descriptionPtr else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Missing store description")
        }
        let description: NSPersistentStoreDescription = cdBorrow(descriptionPtr)
        description.url = cdURL(from: path)
        return CDR_OK
    } catch let error as NSError {
        cdWriteError(error, to: outError)
        return Int32(error.code)
    }
}

@_cdecl("cd_persistent_store_description_get_read_only")
public func cdPersistentStoreDescriptionGetReadOnly(_ descriptionPtr: UnsafeMutableRawPointer?) -> Int32 {
    guard let descriptionPtr else {
        return 0
    }
    let description: NSPersistentStoreDescription = cdBorrow(descriptionPtr)
    return description.isReadOnly ? 1 : 0
}

@_cdecl("cd_persistent_store_description_set_read_only")
public func cdPersistentStoreDescriptionSetReadOnly(
    _ descriptionPtr: UnsafeMutableRawPointer?,
    _ readOnly: Int32,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        guard let descriptionPtr else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Missing store description")
        }
        let description: NSPersistentStoreDescription = cdBorrow(descriptionPtr)
        description.isReadOnly = readOnly != 0
        return CDR_OK
    } catch let error as NSError {
        cdWriteError(error, to: outError)
        return Int32(error.code)
    }
}

@_cdecl("cd_persistent_store_description_get_timeout")
public func cdPersistentStoreDescriptionGetTimeout(_ descriptionPtr: UnsafeMutableRawPointer?) -> Double {
    guard let descriptionPtr else {
        return 0
    }
    let description: NSPersistentStoreDescription = cdBorrow(descriptionPtr)
    return description.timeout
}

@_cdecl("cd_persistent_store_description_set_timeout")
public func cdPersistentStoreDescriptionSetTimeout(
    _ descriptionPtr: UnsafeMutableRawPointer?,
    _ timeout: Double,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        guard let descriptionPtr else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Missing store description")
        }
        let description: NSPersistentStoreDescription = cdBorrow(descriptionPtr)
        description.timeout = timeout
        return CDR_OK
    } catch let error as NSError {
        cdWriteError(error, to: outError)
        return Int32(error.code)
    }
}

@_cdecl("cd_persistent_store_description_sqlite_pragmas_json")
public func cdPersistentStoreDescriptionSQLitePragmasJSON(
    _ descriptionPtr: UnsafeMutableRawPointer?,
    _ outJSON: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        guard let descriptionPtr else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Missing store description")
        }
        let description: NSPersistentStoreDescription = cdBorrow(descriptionPtr)
        let payload = cdStringMapPayload(from: description.sqlitePragmas)
        outJSON?.pointee = cdCString(try cdEncodeJSON(payload))
        return CDR_OK
    } catch let error as NSError {
        cdWriteError(error, to: outError)
        return Int32(error.code)
    }
}

@_cdecl("cd_persistent_store_description_set_sqlite_pragma")
public func cdPersistentStoreDescriptionSetSQLitePragma(
    _ descriptionPtr: UnsafeMutableRawPointer?,
    _ name: UnsafePointer<CChar>?,
    _ value: UnsafePointer<CChar>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        guard let descriptionPtr, let name else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Missing store description or pragma name")
        }
        let description: NSPersistentStoreDescription = cdBorrow(descriptionPtr)
        description.setValue(cdOptionalString(value) as NSString?, forPragmaNamed: String(cString: name))
        return CDR_OK
    } catch let error as NSError {
        cdWriteError(error, to: outError)
        return Int32(error.code)
    }
}

@_cdecl("cd_persistent_store_description_get_should_add_asynchronously")
public func cdPersistentStoreDescriptionGetShouldAddAsynchronously(_ descriptionPtr: UnsafeMutableRawPointer?) -> Int32 {
    guard let descriptionPtr else {
        return 0
    }
    let description: NSPersistentStoreDescription = cdBorrow(descriptionPtr)
    return description.shouldAddStoreAsynchronously ? 1 : 0
}

@_cdecl("cd_persistent_store_description_set_should_add_asynchronously")
public func cdPersistentStoreDescriptionSetShouldAddAsynchronously(_ descriptionPtr: UnsafeMutableRawPointer?, _ asynchronous: Int32) {
    guard let descriptionPtr else {
        return
    }
    let description: NSPersistentStoreDescription = cdBorrow(descriptionPtr)
    description.shouldAddStoreAsynchronously = asynchronous != 0
}

@_cdecl("cd_persistent_store_description_get_should_migrate_automatically")
public func cdPersistentStoreDescriptionGetShouldMigrateAutomatically(_ descriptionPtr: UnsafeMutableRawPointer?) -> Int32 {
    guard let descriptionPtr else {
        return 0
    }
    let description: NSPersistentStoreDescription = cdBorrow(descriptionPtr)
    return description.shouldMigrateStoreAutomatically ? 1 : 0
}

@_cdecl("cd_persistent_store_description_set_should_migrate_automatically")
public func cdPersistentStoreDescriptionSetShouldMigrateAutomatically(_ descriptionPtr: UnsafeMutableRawPointer?, _ automaticallyMigrate: Int32) {
    guard let descriptionPtr else {
        return
    }
    let description: NSPersistentStoreDescription = cdBorrow(descriptionPtr)
    description.shouldMigrateStoreAutomatically = automaticallyMigrate != 0
}

@_cdecl("cd_persistent_store_description_get_should_infer_mapping_model_automatically")
public func cdPersistentStoreDescriptionGetShouldInferMappingModelAutomatically(_ descriptionPtr: UnsafeMutableRawPointer?) -> Int32 {
    guard let descriptionPtr else {
        return 0
    }
    let description: NSPersistentStoreDescription = cdBorrow(descriptionPtr)
    return description.shouldInferMappingModelAutomatically ? 1 : 0
}

@_cdecl("cd_persistent_store_description_set_should_infer_mapping_model_automatically")
public func cdPersistentStoreDescriptionSetShouldInferMappingModelAutomatically(_ descriptionPtr: UnsafeMutableRawPointer?, _ automaticallyInfer: Int32) {
    guard let descriptionPtr else {
        return
    }
    let description: NSPersistentStoreDescription = cdBorrow(descriptionPtr)
    description.shouldInferMappingModelAutomatically = automaticallyInfer != 0
}

@_cdecl("cd_persistent_store_description_set_option_json")
public func cdPersistentStoreDescriptionSetOptionJSON(
    _ descriptionPtr: UnsafeMutableRawPointer?,
    _ key: UnsafePointer<CChar>?,
    _ valueJSON: UnsafePointer<CChar>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        guard let descriptionPtr, let key else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Missing store description or option key")
        }
        let description: NSPersistentStoreDescription = cdBorrow(descriptionPtr)
        let payload = try valueJSON.map { try cdDecodeJSON($0, as: CDValuePayload.self) }
        let foundationValue = payload.flatMap(cdFoundationValue)
        description.setOption(foundationValue as? NSObject, forKey: cdPersistentStoreOptionKey(String(cString: key)))
        return CDR_OK
    } catch let error as NSError {
        cdWriteError(error, to: outError)
        return Int32(error.code)
    }
}

@_cdecl("cd_persistent_container_default_directory")
public func cdPersistentContainerDefaultDirectory() -> UnsafeMutablePointer<CChar>? {
    cdCString(NSPersistentContainer.defaultDirectoryURL().path)
}

@_cdecl("cd_persistent_container_new")
public func cdPersistentContainerNew(
    _ name: UnsafePointer<CChar>?,
    _ modelPtr: UnsafeMutableRawPointer?,
    _ outContainer: UnsafeMutablePointer<UnsafeMutableRawPointer?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        guard let name, let modelPtr else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Missing persistent container name or model")
        }
        let model: NSManagedObjectModel = cdBorrow(modelPtr)
        let container = NSPersistentContainer(name: String(cString: name), managedObjectModel: model)
        outContainer?.pointee = cdRetain(container)
        return CDR_OK
    } catch let error as NSError {
        cdWriteError(error, to: outError)
        return Int32(error.code)
    }
}

@_cdecl("cd_persistent_container_get_name")
public func cdPersistentContainerGetName(_ containerPtr: UnsafeMutableRawPointer?) -> UnsafeMutablePointer<CChar>? {
    guard let containerPtr else {
        return nil
    }
    let container: NSPersistentContainer = cdBorrow(containerPtr)
    return cdCString(container.name)
}

@_cdecl("cd_persistent_container_managed_object_model")
public func cdPersistentContainerManagedObjectModel(_ containerPtr: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let containerPtr else {
        return nil
    }
    let container: NSPersistentContainer = cdBorrow(containerPtr)
    return cdRetain(container.managedObjectModel)
}

@_cdecl("cd_persistent_container_persistent_store_coordinator")
public func cdPersistentContainerPersistentStoreCoordinator(_ containerPtr: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let containerPtr else {
        return nil
    }
    let container: NSPersistentContainer = cdBorrow(containerPtr)
    return cdRetain(container.persistentStoreCoordinator)
}

@_cdecl("cd_persistent_container_persistent_store_descriptions")
public func cdPersistentContainerPersistentStoreDescriptions(_ containerPtr: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let containerPtr else {
        return nil
    }
    let container: NSPersistentContainer = cdBorrow(containerPtr)
    return cdRetain(container.persistentStoreDescriptions as NSArray)
}

@_cdecl("cd_persistent_container_set_persistent_store_descriptions")
public func cdPersistentContainerSetPersistentStoreDescriptions(
    _ containerPtr: UnsafeMutableRawPointer?,
    _ descriptionPtrs: UnsafePointer<UnsafeMutableRawPointer?>?,
    _ count: Int32,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        guard let containerPtr else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Missing persistent container")
        }
        let container: NSPersistentContainer = cdBorrow(containerPtr)
        let descriptions: [NSPersistentStoreDescription] = cdObjects(from: descriptionPtrs, count: count)
        container.persistentStoreDescriptions = descriptions
        return CDR_OK
    } catch let error as NSError {
        cdWriteError(error, to: outError)
        return Int32(error.code)
    }
}

@_cdecl("cd_persistent_container_load_persistent_stores")
public func cdPersistentContainerLoadPersistentStores(
    _ containerPtr: UnsafeMutableRawPointer?,
    _ timeoutSeconds: Int32,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        guard let containerPtr else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Missing persistent container")
        }
        let container: NSPersistentContainer = cdBorrow(containerPtr)
        let tracker = CDLoadTracker(expectedCount: container.persistentStoreDescriptions.count)
        container.loadPersistentStores { _, error in
            tracker.record(error: error as NSError?)
        }
        if let error = tracker.wait(timeoutSeconds: timeoutSeconds) {
            throw error
        }
        return CDR_OK
    } catch let error as NSError {
        cdWriteError(error, to: outError)
        return Int32(error.code)
    }
}

@_cdecl("cd_persistent_container_view_context")
public func cdPersistentContainerViewContext(_ containerPtr: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let containerPtr else {
        return nil
    }
    let container: NSPersistentContainer = cdBorrow(containerPtr)
    return cdRetain(container.viewContext)
}

@_cdecl("cd_persistent_container_new_background_context")
public func cdPersistentContainerNewBackgroundContext(_ containerPtr: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let containerPtr else {
        return nil
    }
    let container: NSPersistentContainer = cdBorrow(containerPtr)
    return cdRetain(container.newBackgroundContext())
}

File: src/persistent_container.rs
use std::collections::BTreeMap;
use std::path::{Path, PathBuf};

use crate::error::CoreDataError;
use crate::ffi;
use crate::model::NSManagedObjectModel;
use crate::private::{
    collect_array, cstring_from_str, error_from_status, impl_object_wrapper, json_cstring,
    opt_cstring_ptr, parse_json_ptr, path_cstring, take_string,
};
use crate::store::NSPersistentContainer;
use crate::store::NSPersistentStoreCoordinator;
use crate::value::{Value, ValuePayload};

pub mod option_keys {
    pub const READ_ONLY: &str = "NSReadOnlyPersistentStoreOption";
    pub const MIGRATE_PERSISTENT_STORES_AUTOMATICALLY: &str =
        "NSMigratePersistentStoresAutomaticallyOption";
    pub const INFER_MAPPING_MODEL_AUTOMATICALLY: &str = "NSInferMappingModelAutomaticallyOption";
    pub const PERSISTENT_HISTORY_TRACKING: &str = "NSPersistentHistoryTrackingKey";
    pub const REMOTE_CHANGE_NOTIFICATION_POST: &str =
        "NSPersistentStoreRemoteChangeNotificationPostOptionKey";
}

impl_object_wrapper!(NSPersistentStoreDescription);

impl NSPersistentStoreDescription {
    pub fn new() -> Result<Self, CoreDataError> {
        let mut out_description = core::ptr::null_mut();
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_persistent_store_description_new(
                core::ptr::null(),
                &mut out_description,
                &mut out_error,
            )
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        unsafe { Self::from_retained_ptr(out_description, "persistent store description") }
    }

    pub fn with_url(path: impl AsRef<Path>) -> Result<Self, CoreDataError> {
        let path = path_cstring(path.as_ref(), "persistent store description URL")?;
        let mut out_description = core::ptr::null_mut();
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_persistent_store_description_new(
                path.as_ptr(),
                &mut out_description,
                &mut out_error,
            )
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        unsafe { Self::from_retained_ptr(out_description, "persistent store description") }
    }

    pub fn store_type(&self) -> Result<String, CoreDataError> {
        let ptr = unsafe { ffi::cd_persistent_store_description_get_type(self.as_ptr()) };
        unsafe { take_string(ptr) }
            .ok_or_else(|| CoreDataError::bridge(-1, "persistent store description type was nil"))
    }

    pub fn set_store_type(&self, store_type: &str) -> Result<(), CoreDataError> {
        let store_type = cstring_from_str(store_type, "persistent store description type")?;
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_persistent_store_description_set_type(
                self.as_ptr(),
                store_type.as_ptr(),
                &mut out_error,
            )
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        Ok(())
    }

    pub fn configuration(&self) -> Option<String> {
        let ptr = unsafe { ffi::cd_persistent_store_description_get_configuration(self.as_ptr()) };
        unsafe { take_string(ptr) }
    }

    pub fn set_configuration(&self, configuration: Option<&str>) -> Result<(), CoreDataError> {
        let configuration = configuration
            .map(|value| cstring_from_str(value, "persistent store configuration"))
            .transpose()?;
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_persistent_store_description_set_configuration(
                self.as_ptr(),
                opt_cstring_ptr(&configuration),
                &mut out_error,
            )
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        Ok(())
    }

    pub fn url(&self) -> Option<PathBuf> {
        let ptr = unsafe { ffi::cd_persistent_store_description_get_url(self.as_ptr()) };
        unsafe { take_string(ptr) }.map(PathBuf::from)
    }

    pub fn set_url(&self, path: impl AsRef<Path>) -> Result<(), CoreDataError> {
        let path = path_cstring(path.as_ref(), "persistent store description URL")?;
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_persistent_store_description_set_url(
                self.as_ptr(),
                path.as_ptr(),
                &mut out_error,
            )
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        Ok(())
    }

    pub fn clear_url(&self) -> Result<(), CoreDataError> {
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_persistent_store_description_set_url(
                self.as_ptr(),
                core::ptr::null(),
                &mut out_error,
            )
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        Ok(())
    }

    pub fn is_read_only(&self) -> bool {
        unsafe { ffi::cd_persistent_store_description_get_read_only(self.as_ptr()) != 0 }
    }

    pub fn set_read_only(&self, read_only: bool) -> Result<(), CoreDataError> {
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_persistent_store_description_set_read_only(
                self.as_ptr(),
                i32::from(read_only),
                &mut out_error,
            )
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        Ok(())
    }

    pub fn timeout(&self) -> f64 {
        unsafe { ffi::cd_persistent_store_description_get_timeout(self.as_ptr()) }
    }

    pub fn set_timeout(&self, timeout: f64) -> Result<(), CoreDataError> {
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_persistent_store_description_set_timeout(self.as_ptr(), timeout, &mut out_error)
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        Ok(())
    }

    pub fn sqlite_pragmas(&self) -> Result<BTreeMap<String, String>, CoreDataError> {
        let mut out_json = core::ptr::null_mut();
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_persistent_store_description_sqlite_pragmas_json(
                self.as_ptr(),
                &mut out_json,
                &mut out_error,
            )
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        unsafe { parse_json_ptr(out_json, "persistent store sqlite pragmas") }
    }

    pub fn set_sqlite_pragma(&self, name: &str, value: Option<&str>) -> Result<(), CoreDataError> {
        let name = cstring_from_str(name, "persistent store sqlite pragma name")?;
        let value = value
            .map(|value| cstring_from_str(value, "persistent store sqlite pragma value"))
            .transpose()?;
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_persistent_store_description_set_sqlite_pragma(
                self.as_ptr(),
                name.as_ptr(),
                opt_cstring_ptr(&value),
                &mut out_error,
            )
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        Ok(())
    }

    pub fn should_add_asynchronously(&self) -> bool {
        unsafe {
            ffi::cd_persistent_store_description_get_should_add_asynchronously(self.as_ptr()) != 0
        }
    }

    pub fn set_should_add_asynchronously(&self, asynchronous: bool) {
        unsafe {
            ffi::cd_persistent_store_description_set_should_add_asynchronously(
                self.as_ptr(),
                i32::from(asynchronous),
            );
        }
    }

    pub fn should_migrate_automatically(&self) -> bool {
        unsafe {
            ffi::cd_persistent_store_description_get_should_migrate_automatically(self.as_ptr())
                != 0
        }
    }

    pub fn set_should_migrate_automatically(&self, automatically_migrate: bool) {
        unsafe {
            ffi::cd_persistent_store_description_set_should_migrate_automatically(
                self.as_ptr(),
                i32::from(automatically_migrate),
            );
        }
    }

    pub fn should_infer_mapping_model_automatically(&self) -> bool {
        unsafe {
            ffi::cd_persistent_store_description_get_should_infer_mapping_model_automatically(
                self.as_ptr(),
            ) != 0
        }
    }

    pub fn set_should_infer_mapping_model_automatically(&self, automatically_infer: bool) {
        unsafe {
            ffi::cd_persistent_store_description_set_should_infer_mapping_model_automatically(
                self.as_ptr(),
                i32::from(automatically_infer),
            );
        }
    }

    pub fn set_option(&self, key: &str, value: Option<Value>) -> Result<(), CoreDataError> {
        let key = cstring_from_str(key, "persistent store option key")?;
        let value_json = value
            .map(ValuePayload::from)
            .as_ref()
            .map(|payload| json_cstring(payload, "persistent store option value"))
            .transpose()?;
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_persistent_store_description_set_option_json(
                self.as_ptr(),
                key.as_ptr(),
                opt_cstring_ptr(&value_json),
                &mut out_error,
            )
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        Ok(())
    }
}

impl NSPersistentContainer {
    pub fn default_directory() -> Result<PathBuf, CoreDataError> {
        let ptr = unsafe { ffi::cd_persistent_container_default_directory() };
        unsafe { take_string(ptr) }
            .map(PathBuf::from)
            .ok_or_else(|| {
                CoreDataError::bridge(-1, "persistent container default directory was nil")
            })
    }

    pub fn name(&self) -> Result<String, CoreDataError> {
        let ptr = unsafe { ffi::cd_persistent_container_get_name(self.as_ptr()) };
        unsafe { take_string(ptr) }
            .ok_or_else(|| CoreDataError::bridge(-1, "persistent container name was nil"))
    }

    pub fn managed_object_model(&self) -> Result<NSManagedObjectModel, CoreDataError> {
        let ptr = unsafe { ffi::cd_persistent_container_managed_object_model(self.as_ptr()) };
        unsafe { NSManagedObjectModel::from_retained_ptr(ptr, "persistent container model") }
    }

    pub fn persistent_store_coordinator(
        &self,
    ) -> Result<NSPersistentStoreCoordinator, CoreDataError> {
        let ptr =
            unsafe { ffi::cd_persistent_container_persistent_store_coordinator(self.as_ptr()) };
        unsafe {
            NSPersistentStoreCoordinator::from_retained_ptr(
                ptr,
                "persistent container persistent store coordinator",
            )
        }
    }

    pub fn persistent_store_descriptions(
        &self,
    ) -> Result<Vec<NSPersistentStoreDescription>, CoreDataError> {
        let array_ptr =
            unsafe { ffi::cd_persistent_container_persistent_store_descriptions(self.as_ptr()) };
        collect_array(
            array_ptr,
            "persistent container persistent store descriptions",
        )
    }

    pub fn set_persistent_store_descriptions(
        &self,
        descriptions: &[&NSPersistentStoreDescription],
    ) -> Result<(), CoreDataError> {
        let raw_descriptions = descriptions
            .iter()
            .map(|description| description.as_ptr())
            .collect::<Vec<_>>();
        let count = i32::try_from(raw_descriptions.len()).map_err(|_| {
            CoreDataError::bridge(-1, "persistent container description count overflow")
        })?;
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_persistent_container_set_persistent_store_descriptions(
                self.as_ptr(),
                raw_descriptions.as_ptr(),
                count,
                &mut out_error,
            )
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        Ok(())
    }
}
